storage/core/src/server/ServerStorage.ts

```diff
@@ -164,6 +165,102 @@ export abstract class ServerStorage
     targetReference: ObjectReference
   ): Promise<void>;
 
+  private async *listObjectsFiltered(
+    sourceStorage: ServerStorage,
+    sourceDirectory: BaseDirectory,
+    maxPageSize: number,
+    predicate?: (objectReference: ObjectReference) => boolean
+  ): AsyncGenerator<ObjectReference> {
+    for await (const objectPage of sourceStorage.getListObjectsPagedIterator(
+      sourceDirectory,
+      maxPageSize
+    )) {
+      for (const object of objectPage) {
+        if (predicate ? predicate(object) : true) {
+          yield object;
+        }
+      }
+    }
+  }
+
+  /**
+   * Copies objects from a base directory in another {@link ServerStorage} instance to this storage.
+   * @param {ServerStorage} sourceStorage source storage. Must be of the same type as this storage.
+   * @param {BaseDirectory} sourceDirectory base directory in the source storage that will be copied.
+   * @param {BaseDirectory} targetDirectory base directory in the target storage.
+   * @param {Function} predicate optional predicate to filter objects to copy. If not specified, all
+   * objects from the sourceDirectory will be copied.
+   * @returns {Promise<void>}
+   * @note This uses server-side copying. Cross-region copy support depends on the storage provider.
+   */
+  public async copyDirectory(
+    sourceStorage: ServerStorage,
+    sourceDirectory: BaseDirectory,
+    targetDirectory: BaseDirectory,
+    predicate?: (objectReference: ObjectReference) => boolean,
+    copyOptions: CopyOptions = {
+      maxPageSize: 100,
+      maxConcurrency: 50,
+      continueOnError: true,
+    }
+  ): Promise<void> {
+    const taskMap = new Map<string, Promise<string>>();
+    const errors: unknown[] = [];
+    const taskFactory = async (object: ObjectReference) => {
+      const newTaskKey = `${object.baseDirectory}/${object.objectName}`;
+      try {
+        await this.copyObject(sourceStorage, object, {
+          ...object,
+          baseDirectory: targetDirectory.baseDirectory,
+        });
+        return newTaskKey;
+      } catch (error) {
+        throw { key: newTaskKey, error };
+      }
+    };
+
+    const handleSingleTask = async () => {
+      try {
+        const [key] = await Promise.race(taskMap);
+        taskMap.delete(key);
```

Comment:
maybe this could go into `finally`?

Reply:
With finally I would have to declare key before assigning a value to it, which essentially would have more lines of code for not repeating delete call.